libraries/chain/include/eosio/chain/authority_checker.hpp

```diff
@@ -225,19 +225,18 @@ namespace detail {
                      bool r = false;
                      typename permission_cache_type::iterator itr = cached_permissions.end();
 
-                     bool propagate_error = false;
+                     std::invoke_result_t<decltype(checker.permission_to_authority), const permission_level> auth = nullptr;
                      try {
-                        auto&& auth = checker.permission_to_authority( permission.permission );
-                        propagate_error = true;
-                        auto res = cached_permissions.emplace( permission.permission, being_evaluated );
-                        itr = res.first;
-                        r = checker.satisfied( std::forward<decltype(auth)>(auth), cached_permissions, recursion_depth + 1 );
-                     } catch( const permission_query_exception& ) {
-                        if( propagate_error )
-                           throw;
-                        else
-                           return total_weight; // if the permission doesn't exist, continue without it
+                        auth = checker.permission_to_authority( permission.permission );
                      }
+                     catch( const permission_query_exception& ) {}
+                     //permission was either invalid (threw permission_query_exception), or wasn't found
+                     if(!auth)
+                        return total_weight;
```

Comment:
Couldn't you just have added these two lines after line 230 in the original version?

Reply:
In the original 230, `permission_to_authority` calls `get_permission()` which throws when it can't find the permission. So, there is no execution after 230 until the `catch` (which then does the `return total_weight` if `propagate_error` is false).

Reply:
But then, how does this PR reduce the overhead of handling exceptions, if they still occur and we catch them as before?

Reply:
The original code was using `get_permission()`,
https://github.com/AntelopeIO/leap/blob/8a9fa7689aebb0d6df276b8e4c31d3bf64800efe/libraries/chain/authorization_manager.cpp#L259-L263
This throws on either an invalid permission or when chainbase's `get()` throws due to it not existing.

The new code uses `find_permission()`,
https://github.com/AntelopeIO/leap/blob/8a9fa7689aebb0d6df276b8e4c31d3bf64800efe/libraries/chain/authorization_manager.cpp#L253-L257
This only throws on an invalid permission -- chainbase's `find()` just returns a `nullptr` if it can't find the permission.

That's why the new code still handles a `permission_query_exception`,
https://github.com/AntelopeIO/leap/blob/7ab4d4c10c563b291d5f2d3ae2774c696cdf0bf4/libraries/chain/include/eosio/chain/authority_checker.hpp#L228-L232
`permission_to_authority()` can still throw via `find_permission()`'s "Invalid permission" check. I'm not actually sure under what conditions that comes in to play (I've not seen it), but of course any refactor here needs to be exact 1:1 behavior unless we do a significant audit and the previous behavior was to treat that identically to a failed lookup (return `total_weight`).

It does appear like `find_permission()` maybe could be refactored to be `nothrow` by also returning a `nullptr` when an invalid permission is used too. But again, we'd have to investigate all uses of `find_permission()` very closely.

Reply:
I'm just wondering about the benefit of moving  lines 234 to 239 out of the `try` block, which requires the awkward declaration of `auth` on line 228. My apologies if I'm still missing something obvious.

Reply:
> I'm just wondering about the benefit of moving lines 234 to 239 out of the `try` block, which requires the awkward declaration of `auth` on line 228. My apologies if I'm still missing something obvious.

Never mind, got it, thanks Matt for the thorough explanation!